### ISSUE TITLE
Crates Universe: Do I need to replicate intra-workspace `Cargo.toml` dependencies in `BUILD.bazel` files?

### ISSUE THREAD

**Issue by joshparallel** [bazelbuild/rules_rust#1923]
My team has a large Cargo workspace with numerous inter-dependent Cargo subpackages. We primarily use pure Cargo to develop/build code, so `Cargo.toml` files are the source of truth for capturing dependencies, but we'd also like to have the option to build with Bazel, e.g. for Python-Rust bindings (since we use Bazel to run Python code). We'd rather not duplicate all the subpackage intra-workspace dependency info already captured in `Cargo.toml` files into `BUILD.bazel` files, but I can't figure out how to avoid this...? E.g. the `Cargo.Bazel.lock` file appears to only capture external dependency info, and doesn't pick up intra-workspace dependencies, meaning `all_crate_deps()` only returns labels for external dependency targets -- is this expected?

Overall I'm wondering: Is there some way to make `Cargo.toml` files the source of truth for everything, without making/manually populating a duplicative `BUILD.bazel` for every workspace package? Sorry if I'm woefully misunderstanding how the Crates Universe rules work!

**Comment by illicitonion** [comment]
Yes, currently crates_universe only supports external crates, and any in-repo crates need manually written BUILD files. including intra-workspace dependencies.

@Calsign has been working on [a gazelle plugin for Rust](https://github.com/Calsign/gazelle_rust) which would be used to generate your in-repo BUILD file contents - maybe they can give an update on how that's going and how production-ready it is :)

**Comment by Calsign** [comment]
I consider the gazelle plugin largely done for my needs: my personal projects and my day job, although we're not using it scale quite yet.

But these are pure-bazel uses, and currently the plugin doesn't try to maintain compatibility with `Cargo.toml` files. I am interested in adding support for this, but there's some tricky parts. For example, gazelle doesn't make it easy to generate rules for targets in a directory other than the one that the source file is declared in. So if you want your sources to live in `src` (like cargo does), but you want your `BUILD.bazel` to live next to `Cargo.toml` in the parent directory, the current plugin does not support this and it's unclear to me how it should be implemented. If you have ideas, I am open to suggestions and contributions.

**Comment by illicitonion** [comment]
> gazelle doesn't make it easy to generate rules for targets in a directory other than the one that the source file is declared in

FWIW we do this in the Java plugin: https://github.com/bazel-contrib/rules_jvm/blob/main/java/gazelle/generate.go - if running in "module" mode (as defined [here](https://github.com/bazel-contrib/rules_jvm/blob/80780e1b1d9cc041b843acc3155a2828a57d5807/java/gazelle/generate.go#L69) we define one BUILD file at the "module root" for several sub-directories, and ignore the subdirectories. I agree that it's more fiddly (we end up keeping our own map of directory -> list of files so we can do the deferred generation (https://github.com/bazel-contrib/rules_jvm/blob/80780e1b1d9cc041b843acc3155a2828a57d5807/java/gazelle/lang.go#L27-L29)), but doable
